### PR TITLE
feat: support interval comparison

### PR DESCRIPTION
**Pull request by cykbls01**
## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

1. fix timestamp substract timestamp.
2. support interval comparison. From pgsql, 1 month equal to 30 days and 1 day equal to 86400000 ms.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Comment by neverchanje** [comment]
@xiangjinwu PTAL is it the correct behavior of interval?

**Comment by xiangjinwu** [comment]
> @xiangjinwu PTAL is it the correct behavior of interval?

Yes ... and no.

> 1 month equal to 30 days and 1 day equal to 86400000 ms

This is true but only during comparison. PostgreSQL does make this assumption in [its comparison implementation](https://github.com/postgres/postgres/blob/07d683b54af854098cc559d4b8640905f9efa0ea/src/backend/utils/adt/timestamp.c#L2373-L2381).

However, the current `Eq` and `Hash` implementation in this PR violates [this requirement](https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq):
> `k1 == k2 -> hash(k1) == hash(k2)`

TLDR: My suggestion is to implement [`justify_interval`](https://www.postgresql.org/docs/current/functions-datetime.html#:~:text=justify_days%20(%20interval%20)%20%E2%86%92%20interval,23%3A00%3A00) and call it before hashing, as well as before serialize to memcomparable encoding but not value encoding.

The actual problem is: `interval '1' month` compares equal to `interval '30' day`, but they are still distinct values:
```
test=# select interval '1' month = interval '30' day;
 ?column? 
----------
 t
(1 row)

test=# select timestamp '2022-02-01' + interval '1' month;
      ?column?       
---------------------
 2022-03-01 00:00:00
(1 row)

test=# select timestamp '2022-02-01' + interval '30' day;
      ?column?       
---------------------
 2022-03-03 00:00:00
(1 row)
```

This is similar to the issue in `decimal`: `2.1 == 2.10` but they have different meaning scientifically.

**Comment by cykbls01** [comment]
> > @xiangjinwu PTAL is it the correct behavior of interval?
> 
> Yes ... and no.
> 
> > 1 month equal to 30 days and 1 day equal to 86400000 ms
> 
> This is true but only during comparison. PostgreSQL does make this assumption in [its comparison implementation](https://github.com/postgres/postgres/blob/07d683b54af854098cc559d4b8640905f9efa0ea/src/backend/utils/adt/timestamp.c#L2373-L2381).
> 
> However, the current `Eq` and `Hash` implementation in this PR violates [this requirement](https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq):
> 
> > `k1 == k2 -> hash(k1) == hash(k2)`
> 
> TLDR: My suggestion is to implement [`justify_interval`](https://www.postgresql.org/docs/current/functions-datetime.html#:~:text=justify_days%20(%20interval%20)%20%E2%86%92%20interval,23%3A00%3A00) and call it before hashing, as well as before serialize to memcomparable encoding but not value encoding.
> 
> The actual problem is: `interval '1' month` compares equal to `interval '30' day`, but they are still distinct values:
> 
> ```
> test=# select interval '1' month = interval '30' day;
>  ?column? 
> ----------
>  t
> (1 row)
> 
> test=# select timestamp '2022-02-01' + interval '1' month;
>       ?column?       
> ---------------------
>  2022-03-01 00:00:00
> (1 row)
> 
> test=# select timestamp '2022-02-01' + interval '30' day;
>       ?column?       
> ---------------------
>  2022-03-03 00:00:00
> (1 row)
> ```
> 
> This is similar to the issue in `decimal`: `2.1 == 2.10` but they have different meaning scientifically.

Thanks, I add a justify function to make hash value equal.
About actual problem,  since in pgsql timestamp add 1 month or 30 days may have different result. Maybe 30 days equal to 1 month only use in interval operation and when add with timestamp will have other logic.

**Comment by neverchanje** [comment]
Lgtm

**Review by xiangjinwu — APPROVED** [review]
lgtm for the purpose of this PR. But to align with PostgreSQL, `justify_interval` actually converts 30 days to 1 month, rather than 1 month to 30 days.

btw implicit comparison thru the use of memcomparable encoding may still be using the old behavior: compare the 3 fields lexically without justify.
